R/dummy_data.R

```diff
@@ -1330,3 +1330,278 @@ dummy_er <- function(nid = 1000, n_hospitals = 10, time_period = c(2015, 2023),
+    cohort <- as.data.table(cohort)
+    cohort$admission_date_time <- as.POSIXct(cohort$admission_date_time,
+      format = "%Y-%m-%d %H:%M",
+      tz = "UTC"
+    )
+
+    cohort$discharge_date_time <- as.POSIXct(cohort$discharge_date_time,
+      format = "%Y-%m-%d %H:%M",
+      tz = "UTC"
+    )
+
+    cohort$los <- as.numeric(difftime(
+      cohort$discharge_date_time,
+      cohort$admission_date_time,
+      units = "hours"
+    ))
+
+    # generate `df1` based on `cohort`
+    df1 <- generate_id_hospital(cohort = cohort, include_prop = 1, avg_repeats = 4.5, seed = seed)
+
+    nid <- uniqueN(df1$genc_id)
+    n_hospitals <- uniqueN(df1$hospital_num)
+
+    ####### sample the gap between IP admission and ordered date time #######
+    # in days
+    admit_order_gap <- round(rlnorm(nrow(df1), meanlog = 0.817, sdlog = 1.215) - 0.8)
+
+    ####### Set the `ordered_date_time` #######
+    # get ordered date
+    df1[, ordered_date := as.Date(admission_date_time) + ddays(admit_order_gap)]
+
+    # sample ordered time
+    df1[, ordered_time := sample_time_shifted(.N, xi = 7.9, omega = 8.8, alpha = 4.5, min = 4, max = 30, seed = seed)]
+
+    ####### get ordered date time by combining ordered date and time #######
+    df1[, ordered_date_time := ordered_date + dhours(ordered_time)]
+
+    # ensure that `ordered_date_time` is not after `discharge_date_time`
+    # re-sample bad values
+    while (nrow(df1[ordered_date_time >= discharge_date_time, ]) > 0) {
+      df1[
+        ordered_date_time >= discharge_date_time,
+        ordered_date_time := as.Date(admission_date_time) +
+          dhours(sample_time_shifted(.N, xi = 7.9, omega = 8.8, alpha = 4.5, min = 4, max = 30))
+      ]
+    }
+
+    # sample gap between ordered and performed date time, in hours
+    # maximum perform gap is the difference between IP discharge and `ordered_date_time`
+    # this prevents `perform_date_time` from being after `discharge_date_time`
+    df1[, max_perform_gap := as.numeric(
+      difftime(discharge_date_time, ordered_date_time, units = "hours")
+    )]
+
+    df1[, perform_gap := rlnorm_trunc(
+      .N,
+      meanlog = 1.3, sdlog = 1.9, min = 0, max = max_perform_gap
+    )]
+
+    df1[sample(nrow(df1), round(0.06 * nrow(df1))), perform_gap := 0] # set some values to 0
+
+    ####### Get performed date time by adding `perform_gap` to `ordered_date_time` #######
+    df1[, performed_date_time := ordered_date_time + dhours(perform_gap)]
+
+    # performed date time should not be after discharge
+    # re-sample it if performed > discharge was sampled previously
+    df1[performed_date_time > discharge_date_time, performed_date_time := ordered_date_time +
+      dhours(
+        rlnorm_trunc(
+          .N,
+          meanlog = 1.3, sdlog = 1.9, min = 0, max = as.numeric(
+            difftime(discharge_date_time, ordered_date_time, units = "hours")
+          )
+        )
+      )]
```

Comment:
Sorry, one more small suggestion here: For encounters with very long LOS, the function occasionally produces extremely long time differences between ordered date-time and performed date-time. Maybe you could cap performed date-time at a max of 6 months after ordered date-time, by taking the min between the time period possible based on discharged date-time and 6 months? 
```suggestion
    # performed date time should not be after discharge
    # re-sample it if performed > discharge was sampled previously
    df1[performed_date_time > discharge_date_time, performed_date_time := ordered_date_time +
      dhours(
        rlnorm_trunc(
          .N,
          meanlog = 1.3, sdlog = 1.9, min = 0, max = min(as.numeric(
            difftime(discharge_date_time, ordered_date_time, units = "hours"), 6 * 30.4 * 24)
          )
        )
      )]
```

Reply:
Also, when I inspect the distribution of difftimes between performed vs. ordered date-time in hours, it looks a bit wonky. The 0's are imputed, but other than that, the smallest difftime is 23h. Would be worth double-checking what the observed difftimes are in the data (for the types of tests included here) and then adjusting how you sample these).
<img width="925" height="592" alt="Screenshot 2025-11-26 102604" src="https://github.com/user-attachments/assets/18e08ff9-ac8e-4371-8d5d-c9e6ec866573" />